telethon/client/telegrambaseclient.py

```diff
@@ -654,6 +654,21 @@ def set_proxy(self: 'TelegramClient', proxy: typing.Union[tuple, dict]):
             else:
                 connection._proxy = proxy
 
+    def _save_states_and_entities(self: 'TelegramClient'):
+        entities = self._mb_entity_cache.get_all_entities()
+
+        # Piggy-back on an arbitrary TL type with users and chats so the session can understand to read the entities.
+        # It doesn't matter if we put users in the list of chats.
+        self.session.process_entities(types.contacts.ResolvedPeer(None, [e._as_input_peer() for e in entities], []))
+
+        ss, cs = self._message_box.session_state()
+        self.session.set_update_state(0, types.updates.State(**ss, unread_count=0))
+        now = datetime.datetime.now()  # any datetime works; channels don't need it
+        entities_ids = [e.id for e in entities]
+        for channel_id, pts in cs.items():
+            if channel_id in entities_ids:
```

Comment:
This is `O(n)`. While I don't expect the array to be long, it doesn't feel great. It's better if we use a set instead.

```suggestion
        entities_ids = {e.id for e in entities}
        for channel_id, pts in cs.items():
            if channel_id in entities_ids:
```